### DISCUSSION TITLE
Managed Storage for LeechBlockNG

### DISCUSSION THREAD

**Discussion by gelatinbomb** [proginosko/LeechBlockNG#368] [Ideas]
### The Problem
As someone that uses Leechblock a lot I have always faced the same issue over and over again. It is just to easy to bypass the extension by creating a new profile or deleting the actual one. Sure you can force the instalation of the extension through a policy but the extension is clean and does not have any blocking list. This has been a huge problem to me until I found the Managed Storage area.

### The Solution 
Managed Storage allows you to force the load of a configuration file stored as a policy. That means no matter how many profiles you create all of them will have Leechblock with a full configuration set up. It is avaliable on both Chrome and Firefox. It can be a little bit hard to set up as a user but at the end it gives amazing results. 

### Details
Managed Storage requires to store a configuration on the policies.json on chrome and a separate .json file on firefox. Usually you have a "json" property where you have to paste all the configurations of the extension in one single line. Some other extensions like ublock are more fancy and have many properties for each configuration. You can tell on the policy if you want to replace the config files on the first start up or each time the browser loads.

Now this can be hard as I mentioned before, especially for non-tech users. So I would recomend to have a button that generates the json configuration based on the actual preferences on leechblock. Then a small text saying where you have to put the text.

### Some Links that might be usefull
https://github.com/gorhill/uBlock/wiki/Deploying-uBlock-Origin:-configuration
https://www.chromium.org/administrators/configuring-policy-for-extensions/
https://developer.chrome.com/docs/extensions/mv3/manifest/storage/
https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Native_manifests#managed_storage_manifests
https://mozilla.github.io/policy-templates/#3rdparty
https://extensionworkshop.com/documentation/enterprise/adding-policy-support-to-your-extension/

**Comment by proginosko** [comment]
Thanks for this. Very interesting. I will look into it.

**Comment by smayer97** [comment]
OH! This would be a MAJOR boon! I know how to stop the creation of a new profile BUT it is not possible to prevent deleting one, at least not that I have yet found. 

Just FYI, and for anyone reading, I have tried to compile a list of all the extra plugs to limit bypassing Leechblock, here: 
List of FAQs for how to prevent bypassing LeechBlock https://github.com/proginosko/LeechBlockNG/issues/328

Anything that adds to this would be great.

If a solution for this could be implemented, this would plug a big hole.

**Comment by proginosko** [comment]
Added an option to export settings to a JSON file in version 1.6.6. Someone else will need to figure out how to use it for managed storage though...

**Comment by ironhak** [comment]
Hello I'm triying every possible way without success. I'm on linux and I managed to install the extension trough policy at `/etc/firefox/policies/policies.json`:
```
  {
    "policies": {
      "ExtensionSettings": {
        "leechblockng@proginosko.com": {
          "installation_mode": "force_installed",
          "install_url": "https://addons.mozilla.org/firefox/downloads/latest/leechblockng@proginosko.com/latest.xpi",
          "private_browsing": true
        }
      
    }
  }
}
```
This works and prevent the removal / disable of the extension. Then I exported my settings by using the json export button on the GUI of leechblock. I tried using the posts mentioned by OP and creating the file `usr/lib/mozilla/managed-storage/leechblockng@proginosko.com.json`:
```
{
  "name": "leechblockng@proginosko.com",
  "description": "ignored",
  "type": "storage",
  "data": {
    "setName1": "CUSTOM SETTINGS",
    "sites1": "reddit.com youtube.com",
...
```
But this does not work. Does anyone know please how to import the settings trough policies on firefox?
Thanks

**Comment by ironhak** [comment]
I managed to make some progress, placing this file at `/etc/firefox/policies/policies.json`

```
{
  "policies": {
    "ExtensionSettings": {
      "leechblockng@proginosko.com": {
        "installation_mode": "force_installed",
        "install_url": "https://addons.mozilla.org/firefox/downloads/latest/leechblockng@proginosko.com/latest.xpi",
        "private_browsing": true
      }
    },

    "3rdparty": {
      "Extensions": {
        "leechblockng@proginosko.com": {
         
    "setName1": "CUSTOM SETTINGS",
    "sites1": "reddit.com youtube.com",
    "times1": "",
    "limitMins1": "",
    "limitPeriod1": "",
/// Rest of config exported from extension button to export json
```

Then as before the extension is isntalled and you can't remove it. Now, the configuration is properly read by Firefox, the problem is that the extension does not read at all from managed storage, but if the extension was reading from it then everything would work fine! And this is the proof: if you go to `about:debugging#/runtime/this-firefox` and you clickl on inspect Leechblock, on the console paste this:
```
const m = await browser.storage.managed.get();
await browser.storage.local.set(m);
console.log("Copied managed -> local");
```

This will copy the settings from managed storage to local storage.... And if you restart firefox you'll see Leechblock finally using the settings!!!! 

Now it's just a matter of being able to make so that Leechblock actually reads managed storage

**Reply by dummeroncoca** [reply]
@proginosko

**Reply by serv-inc** [reply]
reads from managed storage, but also adjust the UI: grey out the areas that are set by managed storage to indicate they cannot be unset etc. Or are you ok with just setting from managed storage if it exists @proginosko ?

**Comment by dummeroncoca** [comment]
I think this is huge.

**Comment by evavh** [comment]
I have just opened a PR (#753) that adds reading from managed storage :)

**Comment by proginosko** [comment]
Support for managed storage added in version 1.7.3 (available for Firefox now, other browsers to follow soon).

Sorry it took so long!
